contracts/Stabilizer.sol

```diff
@@ -0,0 +1,155 @@
+contract Stabilizer {
+    using SafeMath for uint;
+
+    uint public constant MAX_FEE = 10000;
+    uint public buyFee;
+    uint public sellFee;
+    uint public supplyCap;
+    uint public supply;
+    ERC20 public synth;
+    ERC20 public reserve;
+    address public operator;
+    IStrat public strat;
+    address public governance;
+
+    constructor(ERC20 synth_, ERC20 reserve_, address gov_, uint buyFee_, uint sellFee_, uint supplyCap_) public {
+        require(buyFee_ <= MAX_FEE, "buyFee_ too high");
+        require(sellFee_ <= MAX_FEE, "sellFee_ too high");
+        synth = synth_;
+        reserve = reserve_;
+        governance = gov_;
+        buyFee = buyFee_;
+        sellFee = sellFee_;
+        operator = msg.sender;
+        supplyCap = supplyCap_;
+    }
+
+    modifier onlyOperator {
+        require(msg.sender == operator || msg.sender == governance, "ONLY OPERATOR OR GOV");
+        _;
+    }
+
+    modifier onlyGovernance {
+        require(msg.sender == governance, "ONLY GOV");
+        _;
+    }
+
+    function setOperator(address operator_) public {
+        require(msg.sender == governance || msg.sender == operator, "ONLY GOV OR OPERATOR");
+        operator = operator_;
+    }
+
+    function setBuyFee(uint amount) public onlyOperator {
+        require(amount <= MAX_FEE, "amount too high");
+        buyFee = amount;
+    }
+
+    function setSellFee(uint amount) public onlyOperator {
+        require(amount <= MAX_FEE, "amount too high");
+        sellFee = amount;
+    }
+    
+    function setCap(uint amount) public onlyOperator {
+        supplyCap = amount;
+    }
+
+    function setGovernance(address gov_) public onlyGovernance {
+        governance = gov_;
+    }
+
+    function setStrat(IStrat newStrat) public onlyGovernance {
+        if(address(strat) != address(0)) {
+            uint prevTotalValue = strat.calcTotalValue();
+            strat.divest(prevTotalValue);
+        }
+        reserve.transfer(address(newStrat), reserve.balanceOf(address(this)));
+        newStrat.invest();
+        strat = newStrat;
+    }
+
+    function removeStrat() public onlyGovernance {
+        uint prevTotalValue = strat.calcTotalValue();
+        strat.divest(prevTotalValue);
+
+        strat = IStrat(address(0));
+    }
+
+    function takeProfit() public {
+        uint totalReserves = getTotalReserves();
+        if(totalReserves > supply) {
+            uint profit = totalReserves - supply; // underflow prevented by if condition
+            if(address(strat) != address(0)) {
+                uint bal = reserve.balanceOf(address(this));
+                if(bal < profit) {
+                    strat.divest(profit - bal); // underflow prevented by if condition
+                }
+            }
+            reserve.transfer(governance, profit);
+        }
+    }
+
+    function buy(uint amount) public {
+        if(address(strat) != address(0)) {
+            reserve.transferFrom(msg.sender, address(strat), amount);
```

Comment:
Do we want this to be in the buy? Wouldn't this make trading considerably expensive?

Reply:
Could be triggered every few buys or so or something, idk

Reply:
Depends on the `strat`. Our `YearnV2Strat` does have a buffer functionality for gas savings. Yearn vaults also have their own buffer.